xpdan/tools.py

```diff
@@ -13,89 +13,202 @@
-        upper = mean + threshold
-
-        # single out the too low and too high pixels
-        working_mask *= img > lower[int_r]
-        working_mask *= img < upper[int_r]
-
-        return working_mask.astype(bool)
+    from xpdan.tests.utils import PDFGetterShim as PDFGetter
+from matplotlib.path import Path
+from scipy.sparse import csr_matrix
+from multiprocessing import Pool, cpu_count
+
+from skbeam.core.accumulators.binned_statistic import BinnedStatistic1D
+from skbeam.core.mask import margin, binned_outlier
+
+
+# TODO: speed this up
+def mask_ring(v_list, p_list, a_std=3.):
+    """Find outlier pixels in a single ring
+
+    Parameters
+    ----------
+    v_list: list
+        Values in ring
+    p_list: list
+        Positions in ring
+    a_std: float
+        Acceptable number of standard deviations
+
+    Returns
+    -------
+    list:
+        The positions to be masked
+    """
+    mask_list = []
+    while len(v_list) > 0:
+        norm_v_list = np.abs(v_list - np.mean(v_list)) / np.std(v_list)
+        if np.all(norm_v_list < a_std):
+            break
+        # get the index of the worst pixel
+        worst_idx = np.argmax(norm_v_list)
+        # get the position of the worst pixel
+        worst_p = p_list[worst_idx]
+        # add the worst position to the mask
+        # delete the worst position
+        v_list = np.delete(v_list, worst_idx)
+        p_list = np.delete(p_list, worst_idx)
+        mask_list.append(worst_p)
+    return mask_list
+
+
+# TODO: speed this up
+def new_masking_method(img, geo, alpha=3, tmsk=None):
+    """Sigma Clipping based masking
+
+    Parameters
+    ----------
+    img: np.ndarray
+        The image
+    geo: pyFAI.geometry.Geometry instance
+        The detector geometry information
+    alpha: float
+        The number of standard deviations to clip
+    tmsk: np.ndarray, optional
+        Prior mask. If None don't use a prior mask, defaults to None.
+
+    Returns
+    -------
+    np.ndarray:
+        The mask
+    """
+    r = geo.rArray(img.shape)
+    q = geo.qArray(img.shape) / 10  # type: np.ndarray
+    delta_q = geo.deltaQ(img.shape) / 10  # type: np.ndarray
+
+    pixel_size = [getattr(geo, a) for a in ['pixel1', 'pixel2']]
+    rres = np.hypot(*pixel_size)
+    rbins = np.arange(np.min(r) - rres / 2., np.max(r) + rres / 2., rres / 2.)
+    rbinned = BinnedStatistic1D(r.ravel(), statistic=np.max, bins=rbins)
+
+    qbin_sizes = np.nan_to_num(rbinned(delta_q.ravel()))
+    qbin = np.cumsum(qbin_sizes)
+
+    ipos = np.arange(0, np.size(img)).reshape(img.shape)
+    qbinned = BinnedStatistic1D(q.ravel(), bins=qbin)
+    xy = qbinned.xy
+
+    if tmsk is None:
+        tmsk = np.ones(img.shape)
+
+    ring_values = []
+    ring_positions = []
+
+    # TODO: include tmsk into this to speed things up?
+    for i in np.unique(xy):
+        ring_values.append(img.ravel()[xy == i])
+        ring_positions.append(ipos.ravel()[xy == i])
+    with Pool(cpu_count()) as p:
```

Comment:
on the safe side, at least reserve one cpu for main process
eventually, we might want to move to dask